### DISCUSSION TITLE
MYSTIQUE 240 support

### DISCUSSION THREAD

**Discussion by DeadSurfer** [Nortank12/deepcool-digital-linux#18] [Ideas]
![2024-11-15-180809_hyprshot](https://github.com/user-attachments/assets/c7152a97-7d4d-4af4-a49c-7730f02b95b0)

I have one, but I have no idea how it works or how to send any data to it.
If interested I can help with the tests

Bus 001 Device 008: ID 3633:0009 DeepCool MYSTIQUE

**Comment by Nortank12** [comment]
I love the idea, and I've converted your issue to a discussion, since it is more long-term.

If we could figure out how to send just the basic information to the device, it could be a huge step, since I haven't found any information about the MYSTIQUE series. When I have more time, I'll look up how to send data to an LCD display so we can reverse-engineer your device.

I have also added this discussion to the `README` file, so if anyone has ideas or a solution, they can share it here.

**Comment by rayabanilolly** [comment]
hello @Nortank12 , i love what you did and try to use it on my device mystique 360, but as u mentioned it might not work on mystique series, i do some research how to send to my data to my device, it successful on the first try but when i run my code again i got error timeout and i dont know how to solve it :(

heres my code, i hope it could help

`use rusb::{Context, Error, UsbContext};
use std::time::Duration;

fn main() -> Result<(), Error> {
    // Device VID and PID
    let vid = 0x3633;
    let pid = 0x0009;

    // Initialize the USB context
    let context = Context::new()?;

    // Find the device
    let devices = context.devices()?;
    for device in devices.iter() {
        let descriptor = device.device_descriptor()?;

        if descriptor.vendor_id() == vid && descriptor.product_id() == pid {
            println!("Device Found: VID {:04x}, PID {:04x}", vid, pid);

            // Open the device
            let handle = device.open()?;

            // Claim the interface (usually 0, but verify for your device)
            let interface_number = 0;
            handle.claim_interface(interface_number)?;

            // OUT endpoint address
            let out_endpoint = 0x01;

            // Payload with RGB set to white (0xFF, 0xFF, 0xFF)
            let payload = vec![
                0x1b, 0x00, 0xf0, 0x16, 0xcc, 0x21, 0x01, 0xe6, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00,
                0x09, 0x00, 0x00, 0x02, 0x00, 0x02, 0x00, 0x01, 0x03, 0x40, 0x00, 0x00, 0x00,
                0xff, 0xff, 0xff, // RGB Color (White)
                0xf5, 0xf6, 0xf7, // Additional data (experiment if needed)
            ];

            // Send the data to the OUT endpoint
            handle.write_bulk(out_endpoint, &payload, Duration::from_secs(1))?;
            println!("White color payload sent successfully!");

            return Ok(()); // Exit after successfully sending data
        }
    }

    println!("Device not found!");
    Ok(())
}
`
first try (show usage and temp data on the screen):
![image](https://github.com/user-attachments/assets/6c195a94-f47e-44a5-a124-cf4790463301)

second try (just show deepcool logo on the screen):
![image](https://github.com/user-attachments/assets/2d96e7ec-ad2a-44d3-821f-7d5d5b4c7ee6)

**Reply by Nortank12** [reply]
Hey there! I'm glad to see that you managed to get the device working. I haven't used rusb, but as far as I know, it's a low-level wrapper for libusb, unlike hidapi. What I mean is that you might have to close the connection manually at the end of your code. In hidapi it happens automatically when the device goes out of scope.

I'll look into it this weekend, but I wanted to give you a quick answer. Did you reverse-engineer the packets or find something online? I'm really interested in the source. Maybe I can help you figure out more. Can you send a picture of the working screen as well?

**Reply by rayabanilolly** [reply]
i've tried use hidapi just like you do, but unfortunately i can't see my device listed, so ask chatgpt and its recommend me to use rusb, tbh it's my first time use rust theres a lot of code i dont understand.

im not sure if im reverse engineered the packets, but i am using usbpcap and wireshark to log data transfer from official deepcool software in windows, and ask chatgpt to add RGB color on payload but it doesnt work as expected even though the screen show cpu usage and temp.

i'll send you picture of the working screen tonight.

**Reply by DengZhengyuan** [reply]
Very happy to see some updates! I got a Mystique 360 and am using Ubuntu 24.04, but I have no idea with this program. So, I’m waiting. Thank you, guys, for still being fixing it!

**Reply by rayabanilolly** [reply]
sorry for late update @Nortank12 i dont know why but, when i run code that i embedded last time my mystique 360 not show anything, so i decided to log data from deepcool official app on windows again and copy one of request data to my rust code, this time it show something on the screen

this is req that i copy to my rust code, if only i know more about this data.
![image](https://github.com/user-attachments/assets/4a389f46-e6a0-4e90-a141-a133c24cb26c)

https://github.com/user-attachments/assets/694fddea-6b2b-46e0-8901-d443a640b0fa

**Reply by Nortank12** [reply]
Thanks for the video!

I ran my device with `rusb`, and it had no timeout issues, even after recovering from sleep mode. However, I couldn't run the program with cargo run (I have the udev rule set up for user privilege), but running cargo build and manually starting the program worked every time.

If you look at issue #22, we made huge progress by reverse engineering the source code of the old app. Unfortunately, the new app that controls the MYSTIQUE series has a different source code, which is more difficult to decipher.

I tried to guess how your device works, but it is pretty difficult from one packet. Try to start the app while capturing packets, let it run for a minute, then close the app. After that, export all packets to a file (one per line) and send it here. I'll make a piece of code so you can replay all the packets to see if the device runs as expected.

If it works, log what data is displayed for each packet. I think 5 packets are a good start, and I'll help you analyze them so we can create a mapping table for it. At that point, the device will be ready to be implemented.

**Reply by Nortank12** [reply]
> this is req that i copy to my rust code, if only i know more about this data. ![image](https://private-user-images.githubusercontent.com/29328007/395801351-4a389f46-e6a0-4e90-a141-a133c24cb26c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ1NTUyNzIsIm5iZiI6MTczNDU1NDk3MiwicGF0aCI6Ii8yOTMyODAwNy8zOTU4MDEzNTEtNGEzODlmNDYtZTZhMC00ZTkwLWExNDEtYTEzM2MyNGNiMjZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE4VDIwNDkzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExYmQzZDhlMjkyM2I3ODJlYjI0ZjEwNjMyOTRmYTIyYmJkOTczZDVmMjg3YmNiMGYyZTI0MTc5ZmNjY2Y2NzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Vz3QJMBR4__BYwLmXwgHMw3carhQX1u04RC5E0a_FNY)

Can you send me a picture of a USB_INTERRUPT packet as well?

**Reply by EggyToastwSunnySide** [reply]
I'd really want an update for this

**Reply by ghost** [reply]
> hello @Nortank12 , i love what you did and try to use it on my device mystique 360, but as u mentioned it might not work on mystique series, i do some research how to send to my data to my device, it successful on the first try but when i run my code again i got error timeout and i dont know how to solve it :(
> 
> heres my code, i hope it could help
> 
> `use rusb::{Context, Error, UsbContext}; use std::time::Duration;
> 
> fn main() -> Result<(), Error> { // Device VID and PID let vid = 0x3633; let pid = 0x0009;
> 
> ```
> // Initialize the USB context
> let context = Context::new()?;
> 
> // Find the device
> let devices = context.devices()?;
> for device in devices.iter() {
>     let descriptor = device.device_descriptor()?;
> 
>     if descriptor.vendor_id() == vid && descriptor.product_id() == pid {
>         println!("Device Found: VID {:04x}, PID {:04x}", vid, pid);
> 
>         // Open the device
>         let handle = device.open()?;
> 
>         // Claim the interface (usually 0, but verify for your device)
>         let interface_number = 0;
>         handle.claim_interface(interface_number)?;
> 
>         // OUT endpoint address
>         let out_endpoint = 0x01;
> 
>         // Payload with RGB set to white (0xFF, 0xFF, 0xFF)
>         let payload = vec![
>             0x1b, 0x00, 0xf0, 0x16, 0xcc, 0x21, 0x01, 0xe6, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00,
>             0x09, 0x00, 0x00, 0x02, 0x00, 0x02, 0x00, 0x01, 0x03, 0x40, 0x00, 0x00, 0x00,
>             0xff, 0xff, 0xff, // RGB Color (White)
>             0xf5, 0xf6, 0xf7, // Additional data (experiment if needed)
>         ];
> 
>         // Send the data to the OUT endpoint
>         handle.write_bulk(out_endpoint, &payload, Duration::from_secs(1))?;
>         println!("White color payload sent successfully!");
> 
>         return Ok(()); // Exit after successfully sending data
>     }
> }
> 
> println!("Device not found!");
> Ok(())
> ```
> 
> } ` first try (show usage and temp data on the screen): ![image](https://private-user-images.githubusercontent.com/29328007/394952584-6c195a94-f47e-44a5-a124-cf4790463301.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE0NjA3NDMsIm5iZiI6MTc1MTQ2MDQ0MywicGF0aCI6Ii8yOTMyODAwNy8zOTQ5NTI1ODQtNmMxOTVhOTQtZjQ3ZS00NGE1LWExMjQtY2Y0NzkwNDYzMzAxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzAyVDEyNDcyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk3YzYzZmVhYzIwNTljN2JkZjgwNGRhNzI1MjgwMTI0YmJhM2FjZGE2M2M3ZjRkNjcxYjJmNGYxMGQ2ZDdjOTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.coX0oLGpP1ijFVoWAJERgJr0ZJF5RFvSqewNVqyaes8)
> 
> second try (just show deepcool logo on the screen): ![image](https://private-user-images.githubusercontent.com/29328007/394952822-2d96e7ec-ad2a-44d3-821f-7d5d5b4c7ee6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE0NjA3NDMsIm5iZiI6MTc1MTQ2MDQ0MywicGF0aCI6Ii8yOTMyODAwNy8zOTQ5NTI4MjItMmQ5NmU3ZWMtYWQyYS00NGQzLTgyMWYtN2Q1ZDViNGM3ZWU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzAyVDEyNDcyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYzNzc2OTFiZmIwZTk4ZmJhNTM0MmU5NDExOTliY2VlNDYyOTE3NjAxMWIxZmRkYzE4MGFmNmE2MDgxODAzMGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.PiTv9Skq31YxRw1bu8E2Qj6hDsVpO-kx4-yt0iy_ylA)

When i list my devices my list is empty

**Reply by ghost** [reply]
> hello @Nortank12 , i love what you did and try to use it on my device mystique 360, but as u mentioned it might not work on mystique series, i do some research how to send to my data to my device, it successful on the first try but when i run my code again i got error timeout and i dont know how to solve it :(
> 
> heres my code, i hope it could help
> 
> `use rusb::{Context, Error, UsbContext}; use std::time::Duration;
> 
> fn main() -> Result<(), Error> { // Device VID and PID let vid = 0x3633; let pid = 0x0009;
> 
> ```
> // Initialize the USB context
> let context = Context::new()?;
> 
> // Find the device
> let devices = context.devices()?;
> for device in devices.iter() {
>     let descriptor = device.device_descriptor()?;
> 
>     if descriptor.vendor_id() == vid && descriptor.product_id() == pid {
>         println!("Device Found: VID {:04x}, PID {:04x}", vid, pid);
> 
>         // Open the device
>         let handle = device.open()?;
> 
>         // Claim the interface (usually 0, but verify for your device)
>         let interface_number = 0;
>         handle.claim_interface(interface_number)?;
> 
>         // OUT endpoint address
>         let out_endpoint = 0x01;
> 
>         // Payload with RGB set to white (0xFF, 0xFF, 0xFF)
>         let payload = vec![
>             0x1b, 0x00, 0xf0, 0x16, 0xcc, 0x21, 0x01, 0xe6, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00,
>             0x09, 0x00, 0x00, 0x02, 0x00, 0x02, 0x00, 0x01, 0x03, 0x40, 0x00, 0x00, 0x00,
>             0xff, 0xff, 0xff, // RGB Color (White)
>             0xf5, 0xf6, 0xf7, // Additional data (experiment if needed)
>         ];
> 
>         // Send the data to the OUT endpoint
>         handle.write_bulk(out_endpoint, &payload, Duration::from_secs(1))?;
>         println!("White color payload sent successfully!");
> 
>         return Ok(()); // Exit after successfully sending data
>     }
> }
> 
> println!("Device not found!");
> Ok(())
> ```
> 
> } ` first try (show usage and temp data on the screen): ![image](https://private-user-images.githubusercontent.com/29328007/394952584-6c195a94-f47e-44a5-a124-cf4790463301.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE4ODMwMTksIm5iZiI6MTc1MTg4MjcxOSwicGF0aCI6Ii8yOTMyODAwNy8zOTQ5NTI1ODQtNmMxOTVhOTQtZjQ3ZS00NGE1LWExMjQtY2Y0NzkwNDYzMzAxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzA3VDEwMDUxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ5YmY0YjEzM2MwMjI5Y2U1NWM1ZjAwYzBlNDRhMjA2MzY1ZjFmNmYzY2JlMmIyYWI5MzY3ODVkZGIwMjQyODQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.7Ma3HBGzaxmmpkfqTSkVR2mYvspg7_yZIrmN6Mn3pq4)
> 
> second try (just show deepcool logo on the screen): ![image](https://private-user-images.githubusercontent.com/29328007/394952822-2d96e7ec-ad2a-44d3-821f-7d5d5b4c7ee6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE4ODMwMTksIm5iZiI6MTc1MTg4MjcxOSwicGF0aCI6Ii8yOTMyODAwNy8zOTQ5NTI4MjItMmQ5NmU3ZWMtYWQyYS00NGQzLTgyMWYtN2Q1ZDViNGM3ZWU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzA3VDEwMDUxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk1YzE4ZjQ2MTdjMmM4NTM2ZDM5N2ZlYmRiYzQyY2MxMDBjYWIzZDFkODZlY2I0NTZkZDBmNDFlYWJhODFjYWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SQJSFTkLH3O-OMvbH0gxQf49J2lX5b43rTU32RLjJ80)

Can you please get me the full log in winfdws with wireshark for windows startup and mystique running logs

**Comment by dunghoilyd01** [comment]
Are there any update on this device?

**Reply by Nortank12** [reply]
Not yet, as I do not own one, and others got stuck or gave up on testing.

**Reply by ghost** [reply]
i have a mystique 360 running Arch linux i can help you . what you need

**Reply by michelle0812** [reply]
and I have a mystique 360 too, running ubuntu linux, what can i do for you @Nortank12

**Reply by Nortank12** [reply]
If you'd like to reverse engineer the device, you can start by [capturing raw packets in Windows](https://github.com/Nortank12/deepcool-digital-linux/issues/21#issuecomment-2535554199). I've uploaded the [mapping table](https://github.com/Nortank12/deepcool-digital-linux/blob/main/device-list/README.md) of each implemented device (note that the values are in decimal format), but they all have a digital display.

Currently I have a full-time job, but I'll try to decipher their Windows program in the following months; maybe it'll get us closer to the result.

**Comment by HashimSenan** [comment]
since I know nothing of rust I did this in C++ and QT: https://gitlab.com/public_projects5782906/Qt-deepcool-digital. I hobe it helps.

**Comment by jnmcwm** [comment]
> ![2024-11-15-180809_hyprshot](https://private-user-images.githubusercontent.com/41702108/386681974-c7152a97-7d4d-4af4-a49c-7730f02b95b0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjUzMTQ3NzEsIm5iZiI6MTc2NTMxNDQ3MSwicGF0aCI6Ii80MTcwMjEwOC8zODY2ODE5NzQtYzcxNTJhOTctN2Q0ZC00YWY0LWE0OWMtNzczMGYwMmI5NWIwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjA5VDIxMDc1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMwN2JkOGZkZDJkYjVmOGM5MjQ1NWU1YjE5YjgwNGY3NmUyNDQ5N2ZjZjlhZDMzNTFlZTA0ZTA3NzVlMzRkZjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.W1Nq9pdd-zgS9F-NUlqCAHrOmT3zbVKmkd_ZTqvoDxY)
> 
> I have one, but I have no idea how it works or how to send any data to it. If interested I can help with the tests
> 
> Bus 001 Device 008: ID 3633:0009 DeepCool MYSTIQUE

HI, i have the MYSTIQUE 360 and whould also support if i can. But i am a linux newcomer.

**Comment by Herestas** [comment]
Salut tout le monde,  

Deja merci pour le travail  c'est super de pouvoir utiliser l'ecran du boitier sur linux.

Je possede le CH360 DIGITAL

J'ai suivis les instructions dans le ' READ ME'   tout fonctionne bien sauf la partie [OPTIONS]    impossible d'y acceder.

Je vous explique mon probleme  : 

mon écran affiche bien la temperature du CPU, mais pour la partie GPU il reste bloqué sur la vitesse des venntilateurs, donc 0 en IDLE

J'aimerais par conséquents n'afficher que les temperatures du CPU et du GPU  sans la vitesse des ventilos.


C'est agréable de suivre les instructions, meme pour un novice sur linux comme moi



Merci beaucoup     ciao =)

**Comment by mymymy1303** [comment]
Got MYSTIQUE 360 working! After analyzing raw USB packets captured from Windows with Wireshark (and help from vibe coding agents), found that after cold boot you need to send an init sequence on endpoint 0x01 before display data works. Key command is 0x0A with payload EA 07 02 02 02 27 21.

Published working CLI + protocol docs here: https://github.com/mymymy1303/qt-deepcool

**Reply by nerdjb** [reply]
https://github.com/nerdjb/spartacus-control-center

**Comment by nerdjb** [comment]
https://github.com/nerdjb/spartacus-control-center#importing-lcd-wiki-qdt-themes
this is for spartacus

**Comment by LionheartTim** [comment]
Hey guys, it is very basic but i made a version for the mystique. https://github.com/LionheartTim/Deepcool-Mystique-AIO-Linux-with-dashboard
